File: apps/api/app/services/aggregation/objections.py
"""Objection clustering — turns 1,000 free-text objections into ranked themes.

P0 approach: normalize -> exact-key grouping -> greedy token-Jaccard merge.
No embeddings needed locally, deterministic, and O(k^2) only over distinct
keys (small). The inference roadmap swaps this for embedding clustering +
Gemma-27B cluster labeling via the analyst agent.
"""

from __future__ import annotations

from collections import Counter

from ...schemas.reaction import PersonaReaction
from ...schemas.report import ObjectionCluster
from ...utils.text import jaccard, normalize_objection

# Mild prefixes the mock adds for green personas — strip so "minor worry — no
# proof" clusters with "no proof".
_SOFT_PREFIXES = ("only thing i'd check:", "minor worry —", "minor worry -",
                  "before paying i'd still ask:")

# Theme lexicon: different phrasings of the same underlying concern must land
# in ONE cluster ("never says what happens to my data" == "vague on data
# handling"). Order = tie-break priority (more specific themes first).
# Roadmap: replaced by embedding clustering + analyst-agent labeling.
_THEMES: list[tuple[str, set[str]]] = [
    ("security & compliance", {"soc2", "sso", "compliance", "procurement", "security",
                               "pen-test", "audit", "documentation"}),
    ("data privacy", {"data", "privacy", "info", "personal", "handling", "feeding",
                      "private", "stance"}),
    ("free trial & try-before-buy", {"trial", "trying", "test", "paying-first"}),
    ("subscription & lock-in", {"subscription", "subscriptions", "monthly", "billing",
                                "cancel", "recurring", "trap", "lock-in"}),
    ("proof & evidence", {"proof", "evidence", "case", "studies", "study", "numbers",
                          "results", "claims", "adjectives", "prove", "proves",
                          "reference", "manually"}),
    ("AI hype skepticism", {"ai", "ai-powered", "hype", "capability", "bolted",
                            "framing"}),
    ("pricing & affordability", {"price", "pricing", "cost", "costs", "expensive",
                                 "afford", "fees", "budget", "tag", "hidden"}),
    ("setup time & complexity", {"setup", "onboarding", "learn", "migration",
                                 "hours", "system", "buried"}),
    ("integration fit", {"plugs", "stack", "integrate", "integration", "tools",
                         "disconnected"}),
    ("trust & credibility", {"trust", "corporate", "salesy", "oversells", "heard"}),
    ("audience fit", {"solves", "problem", "pitch", "obvious"}),
]


def _clean(text: str) -> str:
    low = text.lower().strip()
    for p in _SOFT_PREFIXES:
        if low.startswith(p):
            return low[len(p):].strip()
    return low


def _theme_of(tokens: set[str]) -> str | None:
    best, best_hits = None, 0
    for name, vocab in _THEMES:
        hits = len(tokens & vocab)
        if hits > best_hits:  # strict > keeps earlier (higher-priority) themes on ties
            best, best_hits = name, hits
    return best


def cluster_objections(reactions: list[PersonaReaction], top_k: int = 8,
                       merge_threshold: float = 0.5) -> list[ObjectionCluster]:
    total = len(reactions) or 1

    # 1) exact grouping on normalized keys
    groups: dict[str, list[PersonaReaction]] = {}
    for r in reactions:
        if not r.first_objection.strip():
            continue
        key = normalize_objection(_clean(r.first_objection))
        groups.setdefault(key, []).append(r)

    # 2a) theme-first merge; 2b) Jaccard fallback for un-themed keys
    themed: dict[str, list[PersonaReaction]] = {}
    merged: list[tuple[set[str], list[PersonaReaction]]] = []
    for key, members in sorted(groups.items(), key=lambda kv: len(kv[1]), reverse=True):
        tokens = set(key.split())
        theme = _theme_of(tokens)
        if theme is not None:
            themed.setdefault(theme, []).extend(members)
            continue
        for existing_tokens, existing_members in merged:
            if jaccard(tokens, existing_tokens) >= merge_threshold:
                existing_members.extend(members)
                existing_tokens |= tokens
                break
        else:
            merged.append((tokens, members))
    merged.extend((set(), members) for members in themed.values())

    # 3) build clusters
    clusters: list[ObjectionCluster] = []
    for _, members in sorted(merged, key=lambda m: len(m[1]), reverse=True)[:top_k]:
        phrasings = Counter(_clean(m.first_objection) for m in members)
        label = phrasings.most_common(1)[0][0]
        # prefer a quote from an actual detractor as the example
        detractors = [m for m in members if m.status in ("red", "yellow")]
        example = (detractors[0] if detractors else members[0]).quote
        seg_counts = Counter(m.segment for m in members if m.segment)
        clusters.append(ObjectionCluster(
            label=label,
            count=len(members),
            share=round(len(members) / total, 4),
            example_quote=example,
            top_segments=[s for s, _ in seg_counts.most_common(2)],
        ))
    return clusters

File: apps/web/lib/server/engine/aggregation/objections.ts
/**
 * Objection clustering — port of
 * apps/api/app/services/aggregation/objections.py.
 * normalize -> exact-key grouping -> theme-first + greedy token-Jaccard merge.
 */

import { jaccard, mostCommon, normalizeObjection, round } from "../text";
import type { PersonaReaction } from "../types";
import type { ObjectionCluster } from "../report";

const SOFT_PREFIXES = ["only thing i'd check:", "minor worry —", "minor worry -", "before paying i'd still ask:"];

const THEMES: [string, Set<string>][] = [
  ["security & compliance", new Set(["soc2", "sso", "compliance", "procurement", "security", "pen-test", "audit", "documentation"])],
  ["data privacy", new Set(["data", "privacy", "info", "personal", "handling", "feeding", "private", "stance"])],
  ["free trial & try-before-buy", new Set(["trial", "trying", "test", "paying-first"])],
  ["subscription & lock-in", new Set(["subscription", "subscriptions", "monthly", "billing", "cancel", "recurring", "trap", "lock-in"])],
  ["proof & evidence", new Set(["proof", "evidence", "case", "studies", "study", "numbers", "results", "claims", "adjectives", "prove", "proves", "reference", "manually"])],
  ["AI hype skepticism", new Set(["ai", "ai-powered", "hype", "capability", "bolted", "framing"])],
  ["pricing & affordability", new Set(["price", "pricing", "cost", "costs", "expensive", "afford", "fees", "budget", "tag", "hidden"])],
  ["setup time & complexity", new Set(["setup", "onboarding", "learn", "migration", "hours", "system", "buried"])],
  ["integration fit", new Set(["plugs", "stack", "integrate", "integration", "tools", "disconnected"])],
  ["trust & credibility", new Set(["trust", "corporate", "salesy", "oversells", "heard"])],
  ["audience fit", new Set(["solves", "problem", "pitch", "obvious"])],
];

function clean(text: string): string {
  const low = text.toLowerCase().trim();
  for (const p of SOFT_PREFIXES) {
    if (low.startsWith(p)) return low.slice(p.length).trim();
  }
  return low;
}

function themeOf(tokens: Set<string>): string | null {
  let best: string | null = null;
  let bestHits = 0;
  for (const [name, vocab] of THEMES) {
    let hits = 0;
    for (const t of tokens) if (vocab.has(t)) hits += 1;
    if (hits > bestHits) {
      best = name;
      bestHits = hits;
    }
  }
  return best;
}

export function clusterObjections(reactions: PersonaReaction[], topK = 8, mergeThreshold = 0.5): ObjectionCluster[] {
  const total = reactions.length || 1;

  // 1) exact grouping on normalized keys (preserve first-seen order).
  const groups = new Map<string, PersonaReaction[]>();
  for (const r of reactions) {
    if (!r.first_objection.trim()) continue;
    const key = normalizeObjection(clean(r.first_objection));
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key)!.push(r);
  }

  // 2) theme-first merge; Jaccard fallback for un-themed keys.
  const themed = new Map<string, PersonaReaction[]>();
  const merged: { tokens: Set<string>; members: PersonaReaction[] }[] = [];
  const sortedGroups = Array.from(groups.entries()).sort((a, b) => b[1].length - a[1].length);
  for (const [key, members] of sortedGroups) {
    const tokens = new Set(key.split(" ").filter(Boolean));
    const theme = themeOf(tokens);
    if (theme !== null) {
      if (!themed.has(theme)) themed.set(theme, []);
      themed.get(theme)!.push(...members);
      continue;
    }
    let placed = false;
    for (const bucket of merged) {
      if (jaccard(tokens, bucket.tokens) >= mergeThreshold) {
        bucket.members.push(...members);
        for (const t of tokens) bucket.tokens.add(t);
        placed = true;
        break;
      }
    }
    if (!placed) merged.push({ tokens, members });
  }
  for (const members of themed.values()) merged.push({ tokens: new Set(), members });

  // 3) build clusters.
  const clusters: ObjectionCluster[] = [];
  const topBuckets = merged.slice().sort((a, b) => b.members.length - a.members.length).slice(0, topK);
  for (const bucket of topBuckets) {
    const phrasings = new Map<string, number>();
    for (const m of bucket.members) {
      const c = clean(m.first_objection);
      phrasings.set(c, (phrasings.get(c) ?? 0) + 1);
    }
    const label = mostCommon(phrasings, 1)[0][0];
    const detractors = bucket.members.filter((m) => m.status === "red" || m.status === "yellow");
    const example = (detractors[0] ?? bucket.members[0]).quote;
    const segCounts = new Map<string, number>();
    for (const m of bucket.members) if (m.segment) segCounts.set(m.segment, (segCounts.get(m.segment) ?? 0) + 1);
    clusters.push({
      label,
      count: bucket.members.length,
      share: round(bucket.members.length / total, 4),
      example_quote: example,
      top_segments: mostCommon(segCounts, 2).map(([s]) => s),
    });
  }
  return clusters;
}
